src/backend/app/auth/auth_routes.py

```diff
@@ -127,45 +127,104 @@ async def logout():
 async def get_or_create_user(
     db: Session,
     user_data: AuthUser,
-) -> DbUser:
+):
     """Get user from User table if exists, else create."""
-    existing_user = await user_crud.get_user(db, user_data.id)
-
-    if existing_user:
-        # Update an existing user
-        if user_data.img_url:
-            existing_user.profile_img = user_data.img_url
-        db.commit()
-        return existing_user
-
-    user_by_username = await user_crud.get_user_by_username(db, user_data.username)
-    if user_by_username:
-        raise HTTPException(
-            status_code=400,
-            detail=(
-                f"User with this username {user_data.username} already exists. "
-                "Please contact the administrator."
-            ),
+    try:
+        update_sql = text(
+            """
+        DO
+        $$
+        BEGIN
+            IF EXISTS (SELECT 1 FROM users WHERE id = :user_id) THEN
+                UPDATE users
+                SET profile_img = :profile_img
+                WHERE id = :user_id;
+            ELSIF EXISTS (SELECT 1 FROM users WHERE username = :username) THEN
+                -- Username already exists, raise an error
+                RAISE EXCEPTION
+                    '
+                    User with this username % already exists
+                    ', :username;
+            ELSE
+                INSERT INTO users (
+                    id, username, profile_img, role, mapping_level,
```

Comment:
This could possibly be simplified by the suggestion Ping made, with an 'upsert' query.

Something like

```
INSERT INTO table_name (column1, column2, ...)
VALUES (value1, value2, ...)
ON CONFLICT (id, username)
DO UPDATE SET column1 = value1, column2 = value2, ...;
```

To make this work do we need a migration to ensure the username field has `UNIQUE` specified? Both changes combined would simplify this query a lot 👍

Reply:
Ok, Sure. By the way, the username field was already a unique field.